New lab on Pod Sandboxing on AKS (#78)
Azure-Samples/aks-labs: docs/security/pod-sandboxing-on-aks.mdx

---
sidebar_position: 1
title: Pod Sandboxing on AKS

> sidebar_label: "Pod Sandboxing"

Reviewer: Do we want this to be the first item under the security section?

Also, can we make the title a little more descriptive, like say "Pod Sandboxing with Kata Containers"

- `--os-sku`: Ensure you select `AzureLinux`, as that is the only OS that supports this feature currently.
- `--node-vm-size`: Please ensure you select a VM size that both supports generation 2 VMs and nested virtualization. The [Dsv3 series](https://learn.microsoft.com/azure/virtual-machines/sizes/general-purpose/dsv3-series?tabs=sizebasic) of VMs are a great example.

:::warning

Pod Sandboxing on AKS is currently in Public Preview, so you will need to register the following feature using Azure CLI.

```bash
az feature register --namespace Microsoft.ContainerService --name KataVMIsolationPreview
```

:::

Run the following command to create an AKS cluster with pod sandboxing enabled.

```bash
az aks create \
--name ${AKS_NAME} \
--resource-group ${RG_NAME} \
--os-sku AzureLinux \
--workload-runtime KataMshvVmIsolation \
--node-vm-size Standard_D4s_v3 \
--node-count 1 \
--generate-ssh-keys
```

Once the AKS cluster has been created, run the following command to connect to it.

```bash
az aks get-credentials \
--resource-group ${RG_NAME} \
--name ${AKS_NAME}
```

### Setup Kubernetes GOAT

> Once your cluster is up and running, navigate to a location that you would like the Kubernetes Goat files to sit at.

Reviewer: Before we have the user go download the K8s Goat files, we should introduce what it is and how they will be using it in the workshop.

> `cd <location_path_of_your_choice>`

Reviewer: If we want users to run a command, we should wrap the code in proper code blocks.

```suggestion

```bash
cd <location_path_of_your_choice>
```
```

Once there, clone the repo:
`git clone https://github.com/madhuakula/kubernetes-goat.git`

> 

Reviewer: ```suggestion
Once there, clone the repo:

```bash
git clone https://github.com/madhuakula/kubernetes-goat.git
```

```

Then navigate into the folder it is installed at:

> `cd kubernetes-goat/`

Reviewer: ```suggestion
Then navigate into the folder it is installed at:

```bash
cd kubernetes-goat
```
```

:::info

Use this as your working directory for the remainder of the workshop.

:::

Kubernetes GOAT offers a [number of scenarios](https://madhuakula.com/kubernetes-goat/docs/scenarios) that one can go through to test their pod's security. For the purposes of this lab, we will only go through a few scenarios that will best illustrate the effect of Pod Sandboxes.

## Privileged Kata Pods

At points in this lab, we will be running some privileged Kata pods. This requires you to [configure your containerd Kata runtime appropriately](https://github.com/kata-containers/kata-containers/blob/main/docs/how-to/privileged.md).

A quick method to do so will be via launching a debug pod. First, use `kubectl get nodes` to get the name of your node, and plug it into the command below:

> `kubectl debug node/<node_name> -it --image=ubuntu:22.04 -- bash`

Reviewer: Rather than telling the user to get the name of the node, we should help them do that.

```suggestion
A quick method to do so will be via launching a debug pod. Run the following commands to get the name of the node then start interactively debugging it:

```bash
NODE_NAME=$(kubectl get nodes -ojsonpath='{.items[0].metadata.name}')
kubectl debug node/$NODE_NAME -it --image=ubuntu:22.04 -- bash
```
```

> Once in your node, you can use an editor of your choice. If you don't have it, you can also quickly install `vim` via `apt update && apt install vim -y`

Reviewer: Instructions below use `vim` so we should just tell the user to use it too. `nano` might be more user friendly too.

```suggestion
Once in your node, you will need to modify files. We will use `vim` for this. Install `vim` using the following command.

```bash
apt update && apt install vim -y
```
```

Then edit your containerd config:

- First, run `vim /host/etc/containerd/config.toml`.
- Look for the section under `[plugins."io.containerd.grpc.v1.cri".containerd.runtimes.kata]`, hit "O" for a new line, and insert `privileged_without_host_devices = true`
- Press "esc", then enter `:wq` to save your changes.

> - Restart containerd ` chroot /host systemctl restart containerd`

Reviewer: Should break these commands up into proper command blocks so that the user can copy/paste into the terminal.

```suggestion
Once vim is installed, edit the containerd config by running the following command:

```bash
vim /host/etc/containerd/config.toml
```

Look for the section under `[plugins."io.containerd.grpc.v1.cri".containerd.runtimes.kata]` (on line 22)

Move your cursor to the end of the line and press the letter `O` on your keyboard to add a new line.

Make sure you indent two spaces, then add the following configuration:

```text
privileged_without_host_devices = true
```

Press the `esc` key on your keyboard, then type `:wq` and press enter to save your changes and exit the file.

Restart containerd with the following command:

```bash
chroot /host systemctl restart containerd
```
```

Reviewer: Should also add a note that the debug session should be terminated at this point.

## Scenarios

### Container Escape

The [Container Escape to the Host System](https://madhuakula.com/kubernetes-goat/docs/scenarios/scenario-4/container-escape-to-the-host-system-in-kubernetes-containers/welcome) scenario illustrates what might happen if permissions that are not required are given to users. In this scenario, you can observe touchpoints that a privileged container might be able to use to impact the host system.

### Non-Kata Demonstration

We will only deploy the specific pod we need for this scenario. In this case, we will first deploy the pods corresponding to the container escape scenario:
`kubectl apply -f scenarios/system-monitor/`

> 

Reviewer: ```suggestion
We will only deploy the specific pod we need for this scenario. In this case, we will first deploy the pods corresponding to the container escape scenario:

```bash
kubectl apply -f scenarios/system-monitor
```

```

Make sure that the pod is up and running. You can do so via `kubectl get pods`.

Once the pod is up and running, we can exec directly into the pod to run some tests.

- Grab your pod name: `kubectl get pods`
- Exec into your pod: `kubectl exec -it <pod-name> -- bash`

> To see the current capabilities of our pod, run `capsh --print`.

Reviewer: ```suggestion
Make sure that the pod is up and running. You can do so using the following command:

```bash
kubectl get pods
```

Next, we can exec directly into the pod to run some tests.

```bash
POD_NAME=$(kubectl get pods -lapp=system-monitor -ojsonpath='{.items[0].metadata.name}')
```

Then exec into your pod: 

```bash
kubectl exec -it $POD_NAME -- bash
```

To see the current capabilities of our pod, run the following command:

```bash
capsh --print
```
```

> At a glance, you should be able to see many capabilities that you would associate with a privileged pod, such as `CAP_SYS_ADMIN`.

Reviewer: I noted it was displayed in lowercase `cap_sys_admin` not uppercase.

> Running `mount` will display the mounted filesystems and their mount points. Take a second to look through the mounted host points. We should see some points that would bring up concerns; a line similar to `http://dev/sda3%20on%20etc/hostname%20type%20ext4%20(rw,relatime)`, for instance, would hint that the host's root filesystem (`/dev/sda`) is mounted at `/host` inside the container, a potential major security vulnerability.

Reviewer: ```suggestion
Also, running the command `mount` will display the mounted filesystems and their mount points. Take a second to look through the mounted host points. We should see some points that would bring up concerns; a line similar to `http://dev/sda3%20on%20etc/hostname%20type%20ext4%20(rw,relatime)`, for instance, would hint that the host's root filesystem (`/dev/sda`) is mounted at `/host` inside the container, a potential major security vulnerability.
```

- You can further explore the mounted filesystem by navigating to the `/host-system/` path via `ls /host-system/`

> Using `chroot`, we can get direct access to host system privileges: `chroot /host-system bash`

Reviewer: ```suggestion
Using `chroot`, we can get direct access to host system privileges: 

```bash
chroot /host-system bash
```
```

> You should now see that you have access to host system resources by running `crictl pods`.

Reviewer: ```suggestion
You should now see that you have access to host system resources by running 

```bash
crictl pods
```
```

> Once we are done here, let's move on to Kata pods. First, let's clean up this deployment. The simplest way would be via `kubectl delete -f scenarios/system-monitor/`.

Reviewer: ```suggestion
Once we are done here, let's move on to Kata pods. First, let's clean up this deployment. Exit the pods interactive terminal then run the following command:

```bash
kubectl delete -f scenarios/system-monitor
```
```

### Kata Demonstration

Now we want to see if running the pod as a Kata pod brings about any differences.

> Let's head over to the deployment YAML, which should be located at `...\kubernetes-goat\scenarios\system-monitor`.

Reviewer: Users should still be in the kubernetes-goat directory.

```suggestion
Let's head over to the deployment YAML, which should be located at `scenarios/system-monitor/deployment.yaml`.
```

Open up the YAML in an editor of your choice.

> In the runtime spec, all you will need to do is add in one line:

Reviewer: ```suggestion
In the deployment template spec (after line 24), add a `runtimeClassName` with a value of `kata-mshv-vm-isolation` as illustrated in the YAML snippet below:
```

> ```YAML

Reviewer: ```suggestion
```yaml
```

spec:
  selector:
    matchLabels:
      app: system-monitor
  template:
    metadata:
      labels:
        app: system-monitor
    spec:
      runtimeClassName: kata-mshv-vm-isolation # one line change to make this pod a Kata sandbox.
```

> Let's deploy the updated pod, and run the same tests as before:
> 
> - We'll first exec into our pod (`kubectl exec -it <pod-name> -- bash`)
> - Run `capsh --print`. This time around, we should notice much less capabilities show up.
> - Run `mount`. Like `capsh`, we should see considerable less mounted resources than before.
> - Let's attempt to get direct access to host system privileges again, and see what host resources we have access to.
>   - We'll first run `chroot /host-system bash`.
>   - Next, we'll run `crictl pods`. This time around, we should be greeted with an error stating that the file/directory does not exist. From the Kata pod, we won't have access to host resources.

Reviewer: ```suggestion
Let's deploy the updated manifest, and run the same tests as before:

Deploy the manifest.

```bash
kubectl apply -f scenarios/system-monitor
```

After you've confirmed the pod is running, exec into it

```bash
POD_NAME=$(kubectl get pods -lapp=system-monitor -ojsonpath='{.items[0].metadata.name}')
kubectl exec -it $POD_NAME -- bash
```

Run the `capsh` command again and this time around, we should notice much less capabilities show up.

```bash
capsh --print
```

Run the `mount` command and again, we should see considerably less mounted resources than before.

```bash
mount
```

Let's attempt to get direct access to host system privileges again and see what host resources we have access to.

We'll first run the `chroot` command again.

```bash
chroot /host-system bash
```

Next, run `crictl` again to view all pods running.

```bash
crictl pods
```

This time around, we should be greeted with an error stating that the file/directory does not exist. From the Kata pod, we won't have access to host resources.
```

With Kata pods, we can see that we generally have less (or in some cases, no) access to resources that we previously had in a non-Kata pod!

> To clean up this first demonstration, run `kubectl delete -f scenarios/system-monitor/` again.

Reviewer: ```suggestion
To clean up this demonstration, exit the interactive terminal session, then run the following command:

```bash
kubectl delete -f scenarios/system-monitor
````
```

## Exploring pod sandboxing under the hood

As mentioned earlier, Pod Sandboxing runs each pods as a virtual machine on top of your node OS.

Let's take a look under the hood to see how Kata pods are set up.

> For this section, let's deploy some sample pods. You can find sample folder for this section under the `docs/getting-started/assets/pod-sandboxing-on-aks` in [AKS labs Github](https://github.com/Azure-Samples/aks-labs/tree/main/docs/security/assets/pod-sandboxing-on-aks/). Please download the folder and place it in the directory you are running the lab on.

Reviewer: We should add very explicit directions on how to download these files instead of leaving it to the user to figure out.

```suggestion
For this section, let's deploy some sample pods. You can find sample folder for this section under the `docs/security/assets/pod-sandboxing-on-aks` in [AKS labs Github](https://github.com/Azure-Samples/aks-labs/tree/main/docs/security/assets/pod-sandboxing-on-aks/). Please download the folder and place it in the directory you are running the lab on.
```

> ```
> kubectl apply -f ./kata_demo/
> ```

Reviewer: In the explicit directions to download the sample files, we need to also explicitly call out that the files should be saved to a folder named `kata_demo`

```suggestion
```bash
kubectl apply -f ./kata_demo/
```
```

1. To list your nodes, use the kubectl get nodes command:

> ```

Reviewer: Good practice to include the type of code block for proper syntax highlighting.

```suggestion
```bash
```

kubectl get nodes -o wide
```

2. Use the kubectl debug command to start a privileged container on your node and connect to it.

> ```
> kubectl debug node/<node-name> -it --image=mcr.microsoft.com/cbl-mariner/busybox:2.0
> ```

Reviewer: Should get the node name for the user. I've left an example of that in a comment above.

```suggestion
```bash
kubectl debug node/<node-name> -it --image=mcr.microsoft.com/cbl-mariner/busybox:2.0
```
```

In the debug pod, lets now navigate to host kernel and take a look at the Kata VMs:

> ```

Reviewer: ```suggestion
```bash
```

chroot /host
ps aux | grep cloud-hypervisor
ps aux | grep containerd-shim-kata
```

> Very interesting! We can see we have a cloud hypervisor running per Kata container within the host. In Pod Sandboxing, Cloud Hypervisor is the default hypervisor for Kata workloads. This differs from upstream Kata Containers, which uses QEMU as a default.

Reviewer: Is there a link to documentation we can leave here?

Lets take a close look at the pods.

> ```

Reviewer: ```suggestion
```bash
```

sudo crictl pods
```

> We can see all our pods that are on the node. We are mainly interested in the Kata pods. Grab the pod ID for your kata pod, and run the command:

Reviewer: Which kata pod? There are two... `kata-1` and `kata-2`

> ```

Reviewer: ```suggestion
```bash
```

sudo crictl ps -a --pod <kata-pod-id>
```

This will provide you with the Kata container id. Lets inspect it further. Grab the container-id from the previous output, and run the command:

> ```

Reviewer: ```suggestion
```bash
```

sudo crictl inspect <container-id> | grep -i runtime
```

> We can see that the runtime is kata based, unlike what you would expect from the normal pod deployments.

Reviewer: ```suggestion
We can see that the runtime is Kata based, unlike what you would expect from the normal pod deployments.
```

Also we should have instruction here to exit the interactive terminal in the node.

## Protecting pods from one another

All compute resources of sandboxed pods are isolated from one another, even if they are on the same node. This can allow you to more confidently bin-pack resources onto one node.

To demonstrate this, we will use the same pods deployed from the previous section. We'll simulate a kernel panic in both a Kata and non-Kata pod, and observe the blast radius.

### Kernel panic - sandboxed

We will crash the sandboxed pod using the SysRq trigger, which simulates a kernel panic inside the sandbox:

> ```

Reviewer: ```suggestion
```bash
```

kubectl exec -it kata-1 -- /bin/sh
echo c > /proc/sysrq-trigger
```

Let's check if other pods are impacted from the kernel panic:

> ```

Reviewer: ```suggestion
```bash
```

kubectl get pods
```

We can see that other pods are alive and well.

### Kernel panic - normal

We now repeat the same experiment on a normal pod:

> ```

Reviewer: ```suggestion
```bash
```

kubectl exec -it normal-1 -- /bin/sh
echo c > /proc/sysrq-trigger
```

Let's check if other pods are impacted from the kernel panic:

> ```

Reviewer: ```suggestion
```bash
```

## Summary

🎉 Congratulations on completing this lab! You should now have some hands-on experience with **Pod Sandboxing** on AKS, with a solid understanding on differences between Kata and normal pods, and how the features of a Kata pod could help with isolating your workloads.

### What we learned

In this lab, you:

- ✅ Set up Pod Sandboxing on AKS.
- ✅ Deployed both sandboxed and non-sandboxed pods on a cluster.
- ✅ Explored isolation provided by Pod Sandboxing.
- ✅ Simulated workload stress scenarios, and saw how Pod Sandboxing can help isolate other workloads from the fallout.

## Next steps

This lab introduced **Pod Sandboxing** for compute isolation on AKS, but there are more concepts you can explore:

- Other [isolation best practices](https://learn.microsoft.com/azure/aks/operator-best-practices-cluster-isolation) on AKS
- Running a thoroughly [multi-tenant setup on AKS](https://learn.microsoft.com/azure/architecture/guide/multitenant/service/aks)

## Cleanup (Optional)

If you are finished with your AKS resources and no longer need them, you can simply delete the resource group to get rid of all components within.

> `az group delete --resource-group myResourceGroup`

Reviewer: ```suggestion

```bash
az group delete --resource-group ${RG_NAME}
```
```

:::warning

Pod Sandboxing on AKS is currently in Public Preview, so you will need to register the following feature using Azure CLI.

```bash
az feature register --namespace Microsoft.ContainerService --name KataVMIsolationPreview
```

:::

Run the following command to create an AKS cluster with pod sandboxing enabled.

```bash
az aks create \
--name ${AKS_NAME} \
--resource-group ${RG_NAME} \
--os-sku AzureLinux \
--workload-runtime KataMshvVmIsolation \
--node-vm-size Standard_D4s_v3 \
--node-count 1 \
--generate-ssh-keys
```

Once the AKS cluster has been created, run the following command to connect to it.

```bash
az aks get-credentials \
--resource-group ${RG_NAME} \
--name ${AKS_NAME}
```

### Setup Kubernetes GOAT

Kubernetes Goat is an intentionally vulnerable Kubernetes cluster designed for learning and practicing Kubernetes security. We will be borrowing sample container images to demonstrate isolation benefits from using pod sandboxing.

Once your cluster is up and running, navigate to a location that you would like the Kubernetes Goat files to sit at.

```bash
cd <location_path_of_your_choice>

> 

Reviewer: ```suggestion
```

```

Once there, clone the repo:

```bash
git clone https://github.com/madhuakula/kubernetes-goat.git`
```

Then navigate into the folder it is installed at:

```bash
cd kubernetes-goat

> 

Reviewer: ```suggestion
```

```

:::info

Use this as your working directory for the remainder of the workshop.

:::

Kubernetes GOAT offers a [number of scenarios](https://madhuakula.com/kubernetes-goat/docs/scenarios) that one can go through to test their pod's security. For the purposes of this lab, we will only go through a few scenarios that will best illustrate the effect of Pod Sandboxes.

## Privileged Kata Pods

At points in this lab, we will be running some privileged Kata pods. This requires you to [configure your containerd Kata runtime appropriately](https://github.com/kata-containers/kata-containers/blob/main/docs/how-to/privileged.md).

A quick method to do so will be via launching a debug pod. Run the following commands to get the name of the node then start interactively debugging it:

```bash
NODE_NAME=$(kubectl get nodes -ojsonpath='{.items[0].metadata.name}')
kubectl debug node/$NODE_NAME -it --image=ubuntu:22.04 -- bash

> 

Reviewer: ```suggestion
```

```

Once in your node, you will need to modify files. We will use `vim` for this. Install `vim` using the following command.

```bash
apt update && apt install vim -y

> 

Reviewer: ```suggestion
```

```

Once vim is installed, edit the containerd config by running the following command:

> ```bash
> vim /host/etc/containerd/config.toml

Reviewer: ```suggestion

```bash
vim /host/etc/containerd/config.toml
```
```

Look for the section under `[plugins."io.containerd.grpc.v1.cri".containerd.runtimes.kata]` (on line 22)

Move your cursor to the end of the line and press the letter O on your keyboard to add a new line.

Make sure you indent two spaces, then add the following configuration:

> ```

Reviewer: ```suggestion
```bash
```

privileged_without_host_devices = true
```

Press the `esc` key on your keyboard, then type `:wq` and press enter to save your changes and exit the file.

Restart containerd with the following command:

> ```

Reviewer: ```suggestion
```bash
```

chroot /host systemctl restart containerd
```

At this point, the debug session should be terminated, and you should be back in your shell.

## Scenarios

### Container Escape

The [Container Escape to the Host System](https://madhuakula.com/kubernetes-goat/docs/scenarios/scenario-4/container-escape-to-the-host-system-in-kubernetes-containers/welcome) scenario illustrates what might happen if permissions that are not required are given to users. In this scenario, you can observe touchpoints that a privileged container might be able to use to impact the host system.

### Non-Kata Demonstration

We will only deploy the specific pod we need for this scenario. In this case, we will first deploy the pods corresponding to the container escape scenario:

> ```bash

Reviewer: ```suggestion

```bash
```

kubectl apply -f scenarios/system-monitor
```

Make sure that the pod is up and running. You can do so using the following command:

> ```bash

Reviewer: ```suggestion

```bash
```

kubectl get pods
```

Next, we can exec directly into the pod to run some tests.

> ```

Reviewer: ```suggestion
```bash
```

POD_NAME=$(kubectl get pods -lapp=system-monitor -ojsonpath='{.items[0].metadata.name}')

> ```

Reviewer: ```suggestion
```

```

Then exec into your pod:

> ```

Reviewer: ```suggestion
```bash
```

kubectl exec -it $POD_NAME -- bash
```

To see the current capabilities of our pod, run the following command:

> ```

Reviewer: ```suggestion

```bash
```

capsh --print
```

At a glance, you should be able to see many capabilities that you would associate with a privileged pod, such as `cap_sys_admin`.

Also, running `mount` will display the mounted filesystems and their mount points. Take a second to look through the mounted host points. We should see some points that would bring up concerns; a line similar to `http://dev/sda3%20on%20etc/hostname%20type%20ext4%20(rw,relatime)`, for instance, would hint that the host's root filesystem (`/dev/sda`) is mounted at `/host` inside the container, a potential major security vulnerability.

- You can further explore the mounted filesystem by navigating to the `/host-system/` path via `ls /host-system/`

Using `chroot`, we can get direct access to host system privileges: 

```bash
chroot /host-system bash

> 

Reviewer: ```suggestion
```

```

You should now see that you have access to host system resources by running 

```bash
crictl pods

> 

Reviewer: ```suggestion
```

```

Once we are done here, let's move on to Kata pods. First, let's clean up this deployment. Exit the pods interactive terminal then run the following command:

```bash
kubectl delete -f scenarios/system-monitor

> 

Reviewer: ```suggestion
```

```

### Kata Demonstration

Now we want to see if running the pod as a Kata pod brings about any differences.

Let's head over to the deployment YAML, which should be located at `scenarios/system-monitor/deployment.yaml`.

Open up the YAML in an editor of your choice.

In the deployment template spec (after line 24), add a `runtimeClassName` with a value of `kata-mshv-vm-isolation` as illustrated in the YAML snippet below:

```yaml
spec:
  selector:
    matchLabels:
      app: system-monitor
  template:
    metadata:
      labels:
        app: system-monitor
    spec:
      runtimeClassName: kata-mshv-vm-isolation # one line change to make this pod a Kata sandbox.

> ```

Reviewer: ```suggestion
```

```

Let's deploy the updated manifest, and run the same tests as before:
Deploy the manifest.

```bash
kubectl apply -f scenarios/system-monitor
```

After you've confirmed the pod is running, exec into it:

```bash
POD_NAME=$(kubectl get pods -lapp=system-monitor -ojsonpath='{.items[0].metadata.name}')
kubectl exec -it $POD_NAME -- bash
```

Run the `capsh` command again and this time around, we should notice much less capabilities show up.

> ```

Reviewer: ```suggestion

```bash
```

capsh --print
```

Run the `mount` command and again, we should see considerably less mounted resources than before.

> ```

Reviewer: ```suggestion

```bash
```

mount
```

Let's attempt to get direct access to host system privileges again and see what host resources we have access to.

We'll first run the `chroot` command again.

> ```

Reviewer: ```suggestion
```bash
```

chroot /host-system bash
```

Next, run `crictl` again to view all pods running.

> ```

Reviewer: ```suggestion
```bash
```

crictl pods
```
This time around, we should be greeted with an error stating that the file/directory does not exist. From the Kata pod, we won't have access to host resources.

With Kata pods, we can see that we generally have less (or in some cases, no) access to resources that we previously had in a non-Kata pod!

To clean up this demonstration, exit the interactive terminal session, then run the following command:

```bash
kubectl delete -f scenarios/system-monitor

> 

Reviewer: ```suggestion
```

```

## Exploring pod sandboxing under the hood

As mentioned earlier, Pod Sandboxing runs each pods as a virtual machine on top of your node OS.

Let's take a look under the hood to see how Kata pods are set up.

For this section, let's deploy some sample pods. You can find sample folder for this section under the `docs/security/assets/pod-sandboxing-on-aks` in [AKS labs Github](https://github.com/Azure-Samples/aks-labs/tree/main/docs/security/assets/pod-sandboxing-on-aks/). Please download the folder and place it in the directory you are running the lab on.

```bash
kubectl apply -f ./kata_demo/

> 

Reviewer: ```suggestion
```

```

1. To list your nodes, use the kubectl get nodes command:

```bash
kubectl get nodes -o wide
```

2. Use the kubectl debug command to start a privileged container on your node and connect to it.

```bash
kubectl debug node/<node-name> -it --image=mcr.microsoft.com/cbl-mariner/busybox:2.0

> 

Reviewer: ```suggestion
```

```

## Summary

🎉 Congratulations on completing this lab! You should now have some hands-on experience with **Pod Sandboxing** on AKS, with a solid understanding on differences between Kata and normal pods, and how the features of a Kata pod could help with isolating your workloads.

### What we learned

In this lab, you:

- ✅ Set up Pod Sandboxing on AKS.
- ✅ Deployed both sandboxed and non-sandboxed pods on a cluster.
- ✅ Explored isolation provided by Pod Sandboxing.
- ✅ Simulated workload stress scenarios, and saw how Pod Sandboxing can help isolate other workloads from the fallout.

## Next steps

This lab introduced **Pod Sandboxing** for compute isolation on AKS, but there are more concepts you can explore:

- Other [isolation best practices](https://learn.microsoft.com/azure/aks/operator-best-practices-cluster-isolation) on AKS
- Running a thoroughly [multi-tenant setup on AKS](https://learn.microsoft.com/azure/architecture/guide/multitenant/service/aks)

## Cleanup (Optional)

If you are finished with your AKS resources and no longer need them, you can simply delete the resource group to get rid of all components within.

> ```bash

Reviewer: ```suggestion

```bash
```
